UpsertSubscriptionToMigrate/handler.ts

```diff
@@ -0,0 +1,268 @@
+        ),
+        TE.fromEither
+      )
+    ),
+    TE.map(ownerId => ({
+      ownerId,
+      subscriptionId
+    }))
+  );
+
+export const getApimUserBySubscription = (
+  config: IDecodableConfigAPIM,
+  apimClient: ApiManagementClient,
+  apimSubscriptionResponse: ApimSubscriptionResponse
+): TE.TaskEither<IApimUserError, ApimUserResponse> =>
+  pipe(
+    TE.tryCatch(
+      () =>
+        apimClient.user.get(
+          config.APIM_RESOURCE_GROUP,
+          config.APIM_SERVICE_NAME,
+          apimSubscriptionResponse.ownerId
+        ),
+      () =>
+        toApimUserError(
+          "The provided subscription identifier is malformed or invalid or occur an Authetication Error."
+        )
+    ),
+    TE.chain(
+      flow(
+        ApimUserResponse.decode,
+        TE.fromEither,
+        TE.mapLeft(() => toApimUserError("Invalid Apim User Response Decode."))
+      )
+    )
+  );
+
+export const mapDataToTableRow = (
+  retrievedDocument: RetrievedService,
+  apimData: {
+    readonly apimUser: ApimDelegateUserResponse;
+    readonly apimSubscription: ApimSubscriptionResponse;
+  }
+): MigrationRowDataTable => ({
+  organizationFiscalCode: retrievedDocument.organizationFiscalCode,
+
+  serviceName: retrievedDocument.serviceName || "",
+  serviceVersion: retrievedDocument.version,
+  sourceEmail: apimData.apimUser.email,
+  sourceId: apimData.apimSubscription.ownerId,
+  sourceName: apimData.apimUser.firstName,
+  sourceSurname: apimData.apimUser.lastName,
+  subscriptionId: retrievedDocument.serviceId
+});
+
+export const queryDataTable = (
+  dbClient: PoolClient,
+  query: string
+): TE.TaskEither<IDbError, QueryResult> =>
+  pipe(
+    TE.tryCatch(
+      () => dbClient.query(query),
+      error => mapPostgreSQLError(error as ErrorPostgreSQL)
+    )
+  );
+
+export const createUpsertSql = (dbConfig: IDecodableConfigPostgreSQL) => (
+  data: MigrationRowDataTable,
+  excludeStatus: "PENDING" = "PENDING"
+): NonEmptyString =>
+  `INSERT INTO "${dbConfig.DB_SCHEMA}"."${dbConfig.DB_TABLE}"(
+        "subscriptionId", "organizationFiscalCode", "sourceId", "sourceName",
+        "sourceSurname", "sourceEmail", "serviceVersion", "serviceName")
+        VALUES ('${data.subscriptionId}', '${data.organizationFiscalCode}', '${data.sourceId}', '${data.sourceName}', '${data.sourceSurname}', '${data.sourceEmail}', '${data.serviceVersion}', '${data.serviceName}')
+        ON CONFLICT ("subscriptionId")
+        DO UPDATE
+            SET "organizationFiscalCode" = "excluded"."organizationFiscalCode",
+            "serviceVersion" = "excluded"."serviceVersion",
+            "serviceName" = "excluded"."serviceName"
+            WHERE "ServicesMigration"."Services"."status" <> '${excludeStatus}'
+            AND "ServicesMigration"."Services"."serviceVersion" < "excluded"."serviceVersion"
+    ` as NonEmptyString;
+
+export const log = (d: unknown): void => {
+  throw new Error(`To be implement ${d}`);
```

Comment:
typo: "to be implemented"

Reply:
the function was actually not used, I remove it in https://github.com/pagopa/io-subscription-migration/pull/15/commits/52c1db54521d2080ef93b49708cca997a9d45be7